### PR TITLE
Fixed #31090 -- Logged transaction management queries.

### PR DESCRIPTION
**Pull request by contextref**
Thanks to @PetterS for the original idea and @felixxm for advice during the DjangoConUS 2022 Sprint!
This addresses https://code.djangoproject.com/ticket/31090

### CODE REVIEW

File: docs/releases/4.2.txt [hunk 1]
```diff
@@ -429,3 +429,6 @@ Miscellaneous
 
 * ``TransactionTestCase.assertQuerysetEqual()`` is deprecated in favor of
   ``assertQuerySetEqual()``.
+
+* The ``django.db.backends.base`` logger now logs transactions commits and 
+  rollbacks at DEBUG level as well as adds them to queries_log
```
Comment (charettes) [inline docs/releases/4.2.txt:434]:
```suggestion
* The ``django.db.backends.base`` logger now logs transaction commits and 
  rollbacks at ``DEBUG`` level.
```

I don't think `queries_log` is a documented interface.

File: tests/backends/base/test_base.py [hunk 2]
```diff
@@ -55,6 +61,55 @@ def test_check_database_version_supported_with_none_as_database_version(self):
...
+    available_apps = []
+
+    @override_settings(DEBUG=True)
+    def test_commit_debug_log(self):
+        conn = connections[DEFAULT_DB_ALIAS]
+        with CaptureQueriesContext(conn):
+            with self.assertLogs("django.db.backends.base", "DEBUG") as cm:
+                with transaction.atomic():
+                    Person.objects.create(first_name="first", last_name="last")
+                    Person.objects.create(first_name="first", last_name="last2")
```
Comment (charettes) [inline tests/backends/base/test_base.py:74]: Creating two objects should not be necessary.

```suggestion
```

File: tests/backends/base/test_base.py [hunk 3]
```diff
@@ -55,6 +61,55 @@ def test_check_database_version_supported_with_none_as_database_version(self):
...
+
+                self.assertEqual(conn.queries_log[-1]["sql"], "COMMIT")
+                self.assertRegex(
+                    cm.output[0],
+                    r"DEBUG:django.db.backends.base:\(\d+.\d{3}\) "
+                    r"COMMIT; alias=default",
```
Comment (charettes) [inline tests/backends/base/test_base.py:80]: I guess we could use `DEFAULT_DB_ALIAS` here as well

```suggestion
                    r"COMMIT; alias=%s" % DEFAULT_DB_ALIAS,
```

File: tests/backends/base/test_base.py [hunk 4]
```diff
@@ -55,6 +61,55 @@ def test_check_database_version_supported_with_none_as_database_version(self):
...
+
+    @override_settings(DEBUG=True)
+    def test_rollback_debug_log(self):
+        conn = connections[DEFAULT_DB_ALIAS]
+        with CaptureQueriesContext(conn):
+            with self.assertLogs("django.db.backends.base", "DEBUG") as cm:
+                with self.assertRaises(Exception):
+                    with transaction.atomic():
+                        Person.objects.create(first_name="first", last_name="last")
+                        Person.objects.create(first_name="first", last_name="last2")
```
Comment (charettes) [inline tests/backends/base/test_base.py:91]: Ditto about the two objects creation

```suggestion
```

File: tests/backends/base/test_base.py [hunk 5]
```diff
@@ -55,6 +61,55 @@ def test_check_database_version_supported_with_none_as_database_version(self):
...
+                    r"DEBUG:django.db.backends.base:\(\d+.\d{3}\) "
+                    r"COMMIT; alias=default",
+                )
+
+    @override_settings(DEBUG=True)
+    def test_rollback_debug_log(self):
+        conn = connections[DEFAULT_DB_ALIAS]
+        with CaptureQueriesContext(conn):
+            with self.assertLogs("django.db.backends.base", "DEBUG") as cm:
+                with self.assertRaises(Exception):
```
Comment (charettes) [inline tests/backends/base/test_base.py:88]: I guess we could use `contextlib.suppress` as well as that's not what we're actually testing. No strong opinion though.

File: tests/backends/base/test_base.py [hunk 6]
```diff
@@ -55,6 +61,55 @@ def test_check_database_version_supported_with_none_as_database_version(self):
...
+                    r"DEBUG:django.db.backends.base:\(\d+.\d{3}\) "
+                    r"ROLLBACK; alias=default",
+                )
+
+    def test_no_logs_without_debug(self):
+        with self.assertNoLogs("django.db.backends.base", "DEBUG"):
+            with self.assertRaises(Exception):
+                with transaction.atomic():
+                    Person.objects.create(first_name="first", last_name="last")
+                    Person.objects.create(first_name="first", last_name="last2")
```
Comment (charettes) [inline tests/backends/base/test_base.py:106]:
```suggestion
```

**Review by charettes — COMMENTED** [review]
Thanks for the contribution @contextref, left a few comments but otherwise the patch looks great!

File: docs/releases/4.2.txt (hunk 1 above)
Comment (contextref) [reply docs/releases/4.2.txt]: good to know, removed!

File: tests/backends/base/test_base.py (hunk 3 above)
Comment (contextref) [reply tests/backends/base/test_base.py]: fixed

File: tests/backends/base/test_base.py (hunk 5 above)
Comment (contextref) [reply tests/backends/base/test_base.py]: yeah, agree that .suppress is better!  fixed.

File: tests/backends/base/test_base.py (hunk 6 above)
Comment (contextref) [reply tests/backends/base/test_base.py]: removed.

File: tests/backends/base/test_base.py (hunk 2 above)
Comment (contextref) [reply tests/backends/base/test_base.py]: removed.

**Comment by contextref** [comment]
> Thanks for the contribution @contextref, left a few comments but otherwise the patch looks great!

thanks for the review @charettes !  Great feedback, I think I addressed everything.

File: django/db/backends/base/base.py [hunk 7]
```diff
@@ -39,6 +39,34 @@ def timezone_constructor(tzname):
...
+                "(%.3f) %s; alias=%s",
+                duration,
+                transaction_end,
+                connection.alias,
+                extra={
+                    "duration": duration,
+                    "sql": transaction_end,
+                    "alias": connection.alias,
+                },
+            )
```
Comment (felixxm) [inline django/db/backends/base/base.py:67]: `CursorDebugWrapper.debug_sql` uses `django.db.backends`. I think we should use the same logger here, update [docs](https://docs.djangoproject.com/en/dev/ref/logging/#django-db-backends):

_"This logging does not include framework-level initialization (e.g. SET TIMEZONE) or transaction management queries (e.g. BEGIN, **COMMIT**, and **ROLLBACK**)."_

and add a `versionchanged` annotation.

**Review by felixxm — COMMENTED** [review]
@contextref Thanks :+1:

File: django/db/backends/base/base.py (hunk 7 above)
Comment (contextref) [reply django/db/backends/base/base.py:67]: cool, thanks for taking a look, I will follow-up within a few days.

Comment (contextref) [reply django/db/backends/base/base.py]: Moved the function to utils.py which allowed to use the proper logger without defining two in base.
Fixed the docs.

Comment (contextref) [reply django/db/backends/base/base.py]: oh, BEGIN is too specific for Postgres.  Removing that assertion.

File: docs/ref/logging.txt [hunk 8]
```diff
@@ -195,11 +195,17 @@ For performance reasons, SQL logging is only enabled when
...
+    not be mentioned.
+
 This logging does not include framework-level initialization (e.g.
-``SET TIMEZONE``) or transaction management queries (e.g. ``BEGIN``,
-``COMMIT``, and ``ROLLBACK``). Turn on query logging in your database if you
+``SET TIMEZONE``). Turn on query logging in your database if you
 wish to view all database queries.
 
+
+
```
Comment (felixxm) [inline docs/ref/logging.txt:208]: Remove blank lines

File: docs/ref/logging.txt [hunk 9]
```diff
@@ -195,11 +195,17 @@ For performance reasons, SQL logging is only enabled when
 ``settings.DEBUG`` is set to ``True``, regardless of the logging
 level or handlers that are installed.
 
+.. versionchanged:: 4.2
+
+    Transaction management queries are now included, so their exemption should
+    not be mentioned.
+
```
Comment (felixxm) [inline docs/ref/logging.txt:202]: This should be move to the end of this section. 
```suggestion
.. versionchanged:: 4.2

    Support for logging transaction management queries (e.g. ``BEGIN``,
    ``COMMIT``, and ``ROLLBACK``) was added.
```

File: django/db/backends/base/base.py (hunk 7 above)
Comment (felixxm) [reply django/db/backends/base/base.py]: `BEGIN` is logged only on SQLite as a workaround to start a transaction explicitly in autocommit mode (see [comment](https://code.djangoproject.com/ticket/31090#comment:4)). I think we should use `debug_transaction_end` to wrap 

https://github.com/django/django/blob/c765b62e3258de4dce9935ab7aed430346dfbc10/django/db/backends/base/base.py#L489-L492

when `autocommit` is `False`.

File: django/db/backends/utils.py [hunk 10]
```diff
@@ -144,6 +144,34 @@ def debug_sql(
             )
 
 
+@contextmanager
+def debug_transaction_end(connection, transaction_end):
```
Comment (felixxm) [inline django/db/backends/utils.py:148]: Maybe:
```suggestion
def debug_transaction(connection, command):
```

Comment (contextref) [reply django/db/backends/utils.py]: done, good idea.

File: docs/ref/logging.txt (hunk 8 above)
Comment (contextref) [reply docs/ref/logging.txt]: done.

File: django/db/backends/base/base.py (hunk 7 above)
Comment (contextref) [reply django/db/backends/base/base.py]: As I understand it you mean wrap the _start_transaction_under_autocommit call?  That is done.

Comment (contextref) [reply django/db/backends/base/base.py]: @felixxm thanks for the feedback, I think I addressed everything, so when you get a chance perhaps you can take another look.

Comment (felixxm) [reply django/db/backends/base/base.py]: `BEGIN` is called explicitly in `_start_transaction_under_autocommit()` so there is no need to log it twice. We should wrap `self._set_autocommit()` when `autocommit` is False as the usual way to start a transaction is to turn `autocommit` off. I pushed edits.

**Comment by felixxm** [comment]
@contextref Thanks :+1: Welcome aboard :boat: 

I pushed tiny edits to docs and fixed `COMMIT` logging on Oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by contextref** [comment]
> @contextref Thanks 👍 Welcome aboard ⛵
> 

Thank you @felixxm for the guidance and the last touch.

**Pull request merged**